### ISSUE TITLE
[REQUEST] Discrete Decision Transformer Implementation

### ISSUE THREAD

**Issue by jdesman1** [takuseno/d3rlpy#327]
**Is your feature request related to a problem? Please describe.**
The d3rlpy library has been excellent for my needs, and I'm grateful for the quick issue resolutions provided by takuseno. Currently, the library supports Decision Transformers only for continuous domains, with discrete functionality noted as planned/under construction.

**Describe the solution you'd like**
Since fragments of the discrete decision transformer implementation [already exist in the codebase](https://github.com/takuseno/d3rlpy/blob/9d4f928fab1311a0f568e6422f36ec50f32e7932/d3rlpy/models/builders.py#L298), could we get an update on the timeline for its full implementation? If it's a sizable task, I'm more than willing to contribute and help bring this feature to fruition.

**Additional context**
N/A

**Comment by takuseno** [comment]
@jdesman1 Thanks for the issue. I'll planning to implement this :smile:

**Comment by jdesman1** [comment]
Hi @takuseno, hope all is well! I wanted to follow up on this since it has been several weeks. Could you please provide an update on the current status of this implementation? I understanding that developing and maintaining d3rlpy is highly time consuming. I see that portions of the code exist already, and if there are tasks that I can contribute to, I can help out. 

Thanks for your time and effort providing such a valuable package!

**Comment by takuseno** [comment]
@jdesman1 In this commit, I've implemented `DiscreteDecisionTransformer`: https://github.com/takuseno/d3rlpy/commit/fcc2789c5040706c9cd09c05bb78e0e6d6711a47 . You can use it via installation from source. Please note that the benchmarking has not been done yet.

**Comment by takuseno** [comment]
I confirmed it works with CartPole at least.
```py
import d3rlpy

dataset, env = d3rlpy.datasets.get_cartpole()

dt = d3rlpy.algos.DiscreteDecisionTransformerConfig(
    batch_size=64,
    num_heads=1,
    learning_rate=1e-4,
    max_timestep=1000,
    num_layers=3,
    position_encoding_type="simple",
    encoder_factory=d3rlpy.models.VectorEncoderFactory([128], exclude_last_activation=True),
    observation_scaler=d3rlpy.preprocessing.StandardObservationScaler(),
    context_size=20,
    warmup_tokens=100000,
).create()

dt.fit(
    dataset,
    n_steps=100000,
    n_steps_per_epoch=1000,
    eval_env=env,
    eval_target_return=500,
)
```

**Comment by jdesman1** [comment]
Thank you for all your hard work! Looking forward to trying this out in the coming days.

**Comment by takuseno** [comment]
@jdesman1 I'm currently benchmarking `DiscreteDecisionTransformer`. After I fixed many details, I confirmed that it performs as well as reported in the paper at least in breakout. I'm collecting scores of the rest of the games now. If you're interested, you can use the latest version in the master branch.
```
$ python reproductions/offline/discrete_decision_transformer.py --game breakout --gpu 0 --pre-stack 
# --pre-stack is optional. If you omit this flag, you can save memory up to 75%. But it slows training speed.
```

**Comment by jdesman1** [comment]
@takuseno Once again, thanks for all your hard work! I've had a chance to use the code a bit and here are my current observations:

# Cartpole
```
import d3rlpy

dataset, env = d3rlpy.datasets.get_cartpole()

dt = d3rlpy.algos.DiscreteDecisionTransformerConfig(
    batch_size=64,
    num_heads=1,
    learning_rate=1e-4,
    max_timestep=1000,
    num_layers=3,
    position_encoding_type="simple",
    encoder_factory=d3rlpy.models.VectorEncoderFactory([128], exclude_last_activation=True),
    observation_scaler=d3rlpy.preprocessing.StandardObservationScaler(),
    context_size=20,
    warmup_tokens=100000,
).create()

dt.fit(
    dataset,
    n_steps=100000,
    n_steps_per_epoch=1000,
    eval_env=env,
    eval_target_return=500,
)
```

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-2-7c1c6ab57e75>](https://localhost:8080/#) in <cell line: 18>()
     16 ).create()
     17 
---> 18 dt.fit(
     19     dataset,
     20     n_steps=100000,

4 frames
[/usr/local/lib/python3.10/dist-packages/d3rlpy/algos/transformer/base.py](https://localhost:8080/#) in fit(self, dataset, n_steps, n_steps_per_epoch, experiment_name, with_timestamp, logger_adapter, show_progress, eval_env, eval_target_return, eval_action_sampler, save_interval, callback)
    275                 dataset.sample_transition().observation_signature.shape[0]
    276             )
--> 277             self.create_impl(observation_shape, action_size)
    278             LOG.debug("Models have been built.")
    279         else:

[/usr/local/lib/python3.10/dist-packages/d3rlpy/base.py](https://localhost:8080/#) in create_impl(self, observation_shape, action_size)
    303         if self._impl:
    304             LOG.warn("Parameters will be reinitialized.")
--> 305         self.inner_create_impl(observation_shape, action_size)
    306 
    307     @abstractmethod

[/usr/local/lib/python3.10/dist-packages/d3rlpy/algos/transformer/decision_transformer.py](https://localhost:8080/#) in inner_create_impl(self, observation_shape, action_size)
    216         self, observation_shape: Shape, action_size: int
    217     ) -> None:
--> 218         transformer = create_discrete_decision_transformer(
    219             observation_shape=observation_shape,
    220             action_size=action_size,

[/usr/local/lib/python3.10/dist-packages/d3rlpy/models/builders.py](https://localhost:8080/#) in create_discrete_decision_transformer(observation_shape, action_size, encoder_factory, num_heads, max_timestep, num_layers, context_size, attn_dropout, resid_dropout, embed_dropout, activation_type, embed_activation_type, position_encoding_type, device)
    330     hidden_size = compute_output_size([observation_shape], encoder)
    331 
--> 332     position_encoding = _create_position_encoding(
    333         position_encoding_type=position_encoding_type,
    334         embed_dim=hidden_size,

[/usr/local/lib/python3.10/dist-packages/d3rlpy/models/builders.py](https://localhost:8080/#) in _create_position_encoding(position_encoding_type, embed_dim, max_timestep, context_size)
    263         )
    264     else:
--> 265         raise ValueError(
    266             f"invalid position_encoding_type: {position_encoding_type}"
    267         )

ValueError: invalid position_encoding_type: simple
```

# CustomDatasets

Training a minimal example works fine:
```
import d3rlpy
import numpy as np

# 1000 steps of observations with shape of (100,)
observations = np.random.random((1000, 100))

# 1000 steps of actions with shape of (4,)
actions = np.random.randint(0, 4, size=(1000,))

# 1000 steps of rewards
rewards = np.random.random(1000)

# 1000 steps of terminal flags
terminals = np.random.randint(2, size=1000)

dataset = d3rlpy.dataset.MDPDataset(
    observations=observations,
    actions=actions,
    rewards=rewards,
    terminals=terminals,
)

dt = d3rlpy.algos.DiscreteDecisionTransformerConfig().create(device='cuda:0')

dt.fit(
    dataset,
    n_steps=200,
    n_steps_per_epoch=100
)

# Error occurs here:
observation = dataset.episodes[0].observations
dt.predict(np.expand_dims(observation, axis=0))
```

```
[/usr/local/lib/python3.10/dist-packages/d3rlpy/algos/transformer/base.py](https://localhost:8080/#) in predict(self, inpt)
    201         assert self._impl is not None, IMPL_NOT_INITIALIZED_ERROR
    202         with torch.no_grad():
--> 203             torch_inpt = TorchTransformerInput.from_numpy(
    204                 inpt=inpt,
    205                 context_size=self._config.context_size,

[/usr/local/lib/python3.10/dist-packages/d3rlpy/algos/transformer/inputs.py](https://localhost:8080/#) in from_numpy(cls, inpt, context_size, device, observation_scaler, action_scaler, reward_scaler)
     59         reward_scaler: Optional[RewardScaler] = None,
     60     ) -> "TorchTransformerInput":
---> 61         if context_size < inpt.length:
     62             observations = slice_observations(
     63                 inpt.observations, inpt.length - context_size, inpt.length

AttributeError: 'Tensor' object has no attribute 'length'
```

# Additional notes
`fit` has no `evaluators` parameter, and implementing the evaluators parameter then requires a `predict_value` function to also be added.

**Comment by takuseno** [comment]
@jdesman1 Thanks for playing this around. First of all, the cartpole example is outdated now. Here is the latest version:
```py
import d3rlpy

dataset, env = d3rlpy.datasets.get_cartpole()

dt = d3rlpy.algos.DiscreteDecisionTransformerConfig(
    batch_size=64,
    num_heads=1,
    learning_rate=1e-4,
    max_timestep=1000,
    num_layers=3,
    position_encoding_type=d3rlpy.PositionEncodingType.SIMPLE,
    encoder_factory=d3rlpy.models.VectorEncoderFactory([128], exclude_last_activation=True),
    observation_scaler=d3rlpy.preprocessing.StandardObservationScaler(),
    context_size=20,
    warmup_tokens=100000,
).create()

dt.fit(
    dataset,
    n_steps=100000,
    n_steps_per_epoch=1000,
    eval_env=env,
    eval_target_return=500,
)
```

Regarding the `predict` method, `DecisionTransformer` needs to be stateful when it interacts with environment. Thus, you need to call `as_stateful_wrapper`:
```py
actor = dt.as_stateful_wrapper(target_return=500)
action = actor.predict(observation, reward)
```
For more details, please check [docs](https://d3rlpy.readthedocs.io/en/v2.1.0/references/algos.html#decision-transformer).

**Comment by jdesman1** [comment]
@takuseno Ah I see, thank you! One quick question: is there any suggested means for OPE of a decision transformer when we don't have access to the original environment for simulation? It seems right now that it is incompatible with the `FQE` classes

**Comment by jdesman1** [comment]
@takuseno Is it possible to yield the predicted returns, possibly enabling the broader compatibility of other OPE methods?

**Comment by takuseno** [comment]
@jdesman1 Sorry for the late response. Currently, Decision Transformer is not compatible with any of Q-learning-based components because DT needs sequence to estimate actions, which requires a little different interface. One thing we can do is to implement a dedicated interface for DT.

**Comment by jdesman1** [comment]
@takuseno That sounds like a great idea! I personally find that DTs work extremely well qualitatively  for some tasks I am working on, but unfortunately it is difficult to quantify since it is not compatible with OPE methods in d3rlpy or outside resources as of now. 

I could see this being of great value. Is there a way I can help implement this?

**Comment by takuseno** [comment]
@jdesman1 Thanks for the offer! The problem is to Decision Transformer to [FQE algorithm](https://github.com/takuseno/d3rlpy/blob/f9bde319d7aeb3f2f5cba983833ed441f19fe0c6/d3rlpy/ope/torch/fqe_impl.py#L30) and create [evaluators](https://github.com/takuseno/d3rlpy/blob/master/d3rlpy/metrics/evaluators.py) compatible with DT. This might take some time, but I'll start looking into it for the next release.

**Comment by takuseno** [comment]
BTW, the `DiscreteDecisionTransformer` has been released in `v2.2.0`. Now, you can use the algorithm without installing d3rlpy from source code.
https://github.com/takuseno/d3rlpy/releases/tag/v2.2.0

**Comment by takuseno** [comment]
Let me close this issue since the discrete version of DT has been released. I'll follow up on the FQE algorithm support.
